Sources/Commands/SwiftTool.swift

```diff
@@ -142,6 +159,31 @@ private class ToolWorkspaceDelegate: WorkspaceDelegate {
         stdoutStream <<< "\n"
         stdoutStream.flush()
     }
+
+    func downloadingBinaryArtifact(from url: String, bytesDownloaded: Int64, totalBytesToDownload: Int64?) {
+        queue.async {
+            if let totalBytesToDownload = totalBytesToDownload {
+                self.downloadProgress[url] = DownloadProgress(
+                    bytesDownloaded: bytesDownloaded,
+                    totalBytesToDownload: totalBytesToDownload)
+            }
+
+            let step = self.downloadProgress.values.reduce(0, { $0 + $1.bytesDownloaded }) / 1024
+            let total = self.downloadProgress.values.reduce(0, { $0 + $1.totalBytesToDownload }) / 1024
+            self.downloadAnimation.update(step: Int(step), total: Int(total), text: "Downloading binary artifacts...")
+        }
+    }
+
+    func didDownloadBinaryArtifacts() {
+        queue.async {
```

Comment:
Do we need to add `queue.sync {}` in the deinit for this class so we don't end up exiting while we have things in the queue?

Reply:
I don't think its necessary. The async closure implicitly captures self, so `ToolWorkspaceDelegate` can't be deinit before all those closures finish running.